src/main/scala/mesosphere/marathon/api/akkahttp/v2/GroupsController.scala

```diff
@@ -39,50 +63,101 @@ class GroupsController(electionService: ElectionService, infoService: GroupInfoS
 
   def appsList(groupId: PathId): Route = ???
 
-  def createGroup(groupId: PathId): Route = ???
+  def createGroup(groupId: PathId)(implicit identity: Identity): Route = {
+    (forceParameter & entity(as(groupUpdateUnmarshaller(groupId)))) { (force, groupUpdate) =>
+      val effectiveGroupId = groupUpdate.id.map(id => PathId(id).canonicalPath(groupId)).getOrElse(groupId)
+      val rootGroup = groupManager.rootGroup()
+
+      if (rootGroup.group(effectiveGroupId).isDefined) { // group already exists
+        reject(Rejections.ConflictingChange(Message(s"Group $effectiveGroupId is already created. Use PUT to change this group.")))
+      } else if (rootGroup.transitiveAppsById.get(effectiveGroupId).isDefined) { // app with the group id already exists
+        reject(Rejections.ConflictingChange(Message(s"An app with the path $effectiveGroupId already exists.")))
+      } else {
+        onComplete(updateOrCreate(groupId, groupUpdate, force)) {
+          case Success(deploymentResult) => complete((StatusCodes.OK, List(Headers.`Marathon-Deployment-Id`(deploymentResult.deploymentId)), deploymentResult))
+          case Failure(ex: IllegalArgumentException) => complete(StatusCodes.UnprocessableEntity -> ex.getMessage)
+          case Failure(ex) => throw ex
+        }
+      }
+    }
+  }
+
+  /**
+    * Until there is async version of updateRoot we have to block here
```

Comment:
I don't quite understand. Why are we blocking?

Reply:
It's a good point